R/data-gather-clean.R

```diff
@@ -4,178 +4,190 @@
-        # If the variable contains "d50", then is most likely ld50 or ed50
-        if (stringr::str_detect(x, "^(ld|ed)50")) {
-          if (as.numeric(last_num) > 50) {
-            # Must be multiple experiments
-            last_num <- stringr::str_replace(last_num, "^50", "")
-          } else {
-            # Single experiment; signal no name change
-            last_num <- NA
-          }
-        }
-      } else {
-        # Must not have digits
-        last_num <- NA
+#' @param data Submission list from JSON
+#' @param section The section name
+#' @param unnest TRUE to unnest the responses; FALSE to have the responses
+#'   nested in a list
+create_table_by_sections <- function(data, section, unnest = FALSE) {
+  # If no data, return NULL
+  if (length(data) == 0) {
+    return(NULL)
+  } else if (length(names(data)) == 1 && names(data) == "experiments") {
+    # If "experiments" is the only element, we need to go deepr to extract
+    # the info for each experiment separately. The section name needs to
+    # have a number to differentiate.
+    dat <- purrr::imap_dfr(
+      data[["experiments"]],
+      function(data, index) {
+        create_table_from_values(
+          data = data,
+          section = section,
+          exp_num = index,
+          unnest = unnest
+        )
       }
-      last_num
-    }
-  )
-  # Append the experiment number to the step
-  data$step <- purrr::map2(data$step, num_list, function(x, y) {
-    if (!is.na(y)) {
-      glue::glue("{x} [{y}]")
-    } else {
-      x
-    }
-  })
-  # Remove the experiment number from the variable
-  data$variable <- purrr::map2(data$variable, num_list, function(x, y) {
-    if (!is.na(y)) {
-      stringr::str_replace(x, glue::glue("{y}$"), "")
-    } else {
-      x
-    }
-  })
-  # Fix list column problem
-  data$step <- as.character(data$step)
-  data$variable <- as.character(data$variable)
-  # No longer need the main_variable and sub_variable columns
-  data <- data[, -which(names(data) %in% c("main_variable", "sub_variable"))]
-  data
+    )
+  } else {
+    # If the section does not contain separate experiments, return the data
+    # from the section
+    dat <- create_table_from_values(
+      data = data,
+      section = section,
+      unnest = unnest
+    )
+  }
+  dat
+}
+
+#' Create a tibble from the values in section
+#'
+#' Create a tibble with section name, experiment number, variables, and
+#' response values.
+#'
+#' @inheritParams create_table_by_sections
+#' @param exp_num Numeric experiment number
+create_table_from_values <- function(data, section, exp_num = NA, unnest) {
+  if (isTRUE(unnest)) {
+    tibble::tibble(
+      section = section,
+      exp_num = exp_num,
+      variable = names(unlist(data)),
+      response = as.character(unlist(data))
+    )
+  } else {
+    tibble::tibble(
+      section = section,
+      exp_num = exp_num,
+      variable = names(data),
+      response = data
+    )
+  }
```

Comment:
Sounds good. I will remove the unnest option. I had left it in case things changed in the future, but there's currently no plans to need it at this time.

Reply:
I actually went through and double checked this, again, when going over the added tests branch. I was thinking it would be nice to have a nested list for things like routes. However, the JSON we get from the submissions is not consistent. Some of the sections are double nested, which completely messes up the table if we do not unnest.

In the end, I removed the unnest option and made it so it always unnests the responses. Somewhat sad since some things (basically just routes) become a smidge less nice to look at in the table, but definitely causes less problems.